### DISCUSSION TITLE
ESP32 OLED Bluetooth Speaker by XPM

### DISCUSSION THREAD

**Discussion by XPMonline** [pschatzmann/ESP32-A2DP#526] [Show and tell]
Hello everyone! When I was in the market looking for a portable bluetooth speaker that would meet my feature requirements, there weren't many options available. So I decided to make one and equip it with top features. And since it's powered by Mr. Schatzmann's library, I'd like to share my experience here.

![20240302_124916](https://github.com/pschatzmann/ESP32-A2DP/assets/161959395/221f3105-b52d-4ddb-97ef-930e4aa23a35)
![20240302_124514](https://github.com/pschatzmann/ESP32-A2DP/assets/161959395/3228713c-e721-4800-abe6-df6d2789fc95)
![20240302_124458](https://github.com/pschatzmann/ESP32-A2DP/assets/161959395/0e238953-575f-4e6b-ba90-7127e4489a0e)
![guts](https://github.com/pschatzmann/ESP32-A2DP/assets/161959395/7be917d3-3188-431f-88b8-8115ad4552de)

Please also check my video https://www.youtube.com/watch?v=3Gfe6xS6U9Q&ab_channel=OneWire

USED COMPONENTS:
- PLA printed body
- ESPRESSIF ESP32 BT/WiFi dual core platform
- 2x5 Watt stereo speakers
- PCM5102 DAC I2S module
- PAM8406 Power Amplifier
- 3000mAh Li-Ion battery
- 1A Li-Ion battery charger and boost converter (TP4056 + boost MT3608 module)
- TEMT6000 ambient light sensor
- Rotary Encoder
- TTP223 capacitive touch sensor
- OLED 1.3" I2C display
- WS2812 RGB Ring module

Things that need to be added/resolved: 
- Support for song time code. I found partial support for av_play_pos_changed(); in the library, but it can't be loaded after boot. For example, the av_playback_changed(); function can be loaded at startup by calling "set_avrc_rn_playstatus_callback", but this option is missing for the "play_pos" function.

- I can't figure out how to play a message like "low battery" while streaming music. I can play the messages from the integrated SPIFFS memory via the "audioI2S" library, but only if no bluetooth music source is connected. If I try to play a message while playing bluetooth music, the platform will reboot.

- Support for additional codecs like aptX.

I am not an experienced programmer, so please excuse my ability to fully understand the code and this library.

If anyone has a good idea how to improve my prototype, or any question, feel free to reply to this topic!

A big thanks to Mr. Phil Schatzmann for a great library!

**Comment by geoplayx** [comment]
Hi! that's what bluetooth speaker should be. Would you like to share the code? thanks!

**Reply by iponkada** [reply]
чтобы давать советы нужно увидеть код. дайте скетч и может смогу помочь.
